LibraBFT/Impl/Consensus/BlockStorage/Properties/BlockTree.agda

```diff
@@ -50,15 +53,108 @@ module insertBlockESpec (block : ExecutedBlock) (bt : BlockTree) where
                 -- between b ^∙ ebBlock ∙ bBlockData and block ^∙ ebBlock ∙
                 -- bBlockdata
 
-      qcPost : ∀ qc → qc QCProps.∈BlockTree bt“ → qc QCProps.∈BlockTree bt ⊎ qc ≡ block ^∙ ebBlock ∙ bQuorumCert
-
+      qcPost : QCProps.∈Post⇒∈PreOrBT (_≡ block ^∙ ebBlock ∙ bQuorumCert) bt“ bt
 
 module insertQuorumCertESpec
   (qc : QuorumCert) (bt0  : BlockTree) where
+  open insertQuorumCertE qc bt0
+
+  Ok : Set
+  Ok = ∃₂ λ bt1 il → insertQuorumCertE qc bt0 ≡ Right (bt1 , il)
+
+  private
+    Ok' : BlockTree → List InfoLog → Either ErrLog (BlockTree × List InfoLog) → Set
+    Ok' bt il m = m ≡ Right (bt , il)
 
   Contract : Either ErrLog (BlockTree × List InfoLog) → Set
   Contract (Left _) = Unit
-  Contract (Right (bt1 , il)) = ∀ {qc'} → qc' ∈BlockTree bt1 → qc' ∈BlockTree bt0 ⊎ qc' ≡ qc
-
-  postulate -- TODO-1: prove it
-    contract : Contract (insertQuorumCertE qc bt0)
+  Contract (Right (bt1 , il)) = ∈Post⇒∈PreOrBT (_≡ qc) bt0 bt1
+                              -- TODO: this is only part of what we need.  We want to also prove
+                              -- that QCs in the pre are also in the post.  Make a record?
+
+  contract : (isOk : Ok) → let (bt1 , il , _) = isOk in Contract (Right (bt1 , il))
+  contract (bt1 , il , isOk)
+     with safetyInvariant
+  ...| Right unit
+    with pf1 isOk
+    where
+    pf1 : Ok' bt1 il (step₁ blockId) → ∃[ block ] (just block ≡ btGetBlock blockId bt0 × Ok' bt1 il (step₂ blockId block))
+    pf1 isOk
+       with  btGetBlock blockId bt0
+    ...| just block = block , refl , isOk
+  ...| block , lookupJust₁ , step2-ok
+    with pf2 step2-ok
+    where
+    pf2 : Ok' bt1 il (step₂ blockId block) → ∃[ hcb ](just hcb ≡ bt0 ^∙ btHighestCertifiedBlock × Ok' bt1 il (step₃ blockId block hcb))
+    pf2 isOk
+       with bt0 ^∙ btHighestCertifiedBlock
+    ...| just hcb   = hcb , refl , isOk
+  ...| hcb , lookupJust₂ , step3-ok
+    with          bt0 & btHighestCertifiedBlockId ∙~ block ^∙ ebId &    btHighestQuorumCert ∙~ qc
+       | inspect (bt0 & btHighestCertifiedBlockId ∙~ block ^∙ ebId &_) (btHighestQuorumCert ∙~ qc)
+       | fakeInfo ∷ []
+  ...| bt₃-true | [ refl ] | il₃-true
+    with  if ⌊ (block ^∙ ebRound) >? (hcb ^∙ ebRound) ⌋ then bt₃-true else   bt0               | inspect
```

Comment:
I think you should try simplifying to just doing analysis on

`(block ^∙ ebRound) >? (hcb ^∙ ebRound)`

and other such expressions. You will then get the equal proof (not just a boolean), which may help you avoid using `inspect`.  The proof state will then move forward once you inspect the result of the decision (i.e., it will be able to take the `then` or `else` branch depending on the result)